video/Multi-dimensional-Sensor-Data

WANGXIAOMIN-HIK opened the pull request:
…roposal, supporting multi-dimensional sensor data as metadata, events, and MQTT reports. Main changes include:

1. **Metadata Stream Support for Sensor Data**:
   - Modified metadatastream.xsd, added SensorDataStream element and SensorData complex type to support XML representation of sensor data.
   - Defined core fields such as SensorType, SensorValue, and SensorUnit, along with necessary attributes and extension points.

2. **Sensor Event Topic Definitions**:
   - Added sensor event topics in Core.xml, defining event topics related to temperature and humidity sensors.
   - Added standard structure descriptions for sensor event messages.
   - Provided XML and JSON examples for sensor events.

3. **MQTT Sensor Data Publication**:
   - Added sensor data MQTT topics and JSON payload definitions in Analytics.xml.
   - Added SensorDataFilter element in event.wsdl to control sensor data topics published via MQTT.

4. **Sensor Capability Descriptions**:
   - Added sensor data related capability description fields in analytics.wsdl, such as SensorDataSupport, SensorTypesSupported, and SensorMQTTSupport.

This PR provides detailed examples for temperature and humidity sensors, laying the foundation for future extensions to other types of sensors.
```

WANGXIAOMIN-HIK committed:
This PR implements the ONVIF Multi-dimensional Sensor Data Standard Proposal, supporting multi-dimensional sensor data as metadata, events, and MQTT reports. Main changes include:

1. **Metadata Stream Support for Sensor Data**:
   - Modified metadatastream.xsd, added SensorDataStream element and SensorData complex type to support XML representation of sensor data.
   - Defined core fields such as SensorType, SensorValue, and SensorUnit, along with necessary attributes and extension points.

2. **Sensor Event Topic Definitions**:
   - Added sensor event topics in Core.xml, defining event topics related to temperature and humidity sensors.
   - Added standard structure descriptions for sensor event messages.
   - Provided XML and JSON examples for sensor events.

3. **MQTT Sensor Data Publication**:
   - Added sensor data MQTT topics and JSON payload definitions in Analytics.xml.
   - Added SensorDataFilter element in event.wsdl to control sensor data topics published via MQTT.

4. **Sensor Capability Descriptions**:
   - Added sensor data related capability description fields in analytics.wsdl, such as SensorDataSupport, SensorTypesSupported, and SensorMQTTSupport.

This PR provides detailed examples for temperature and humidity sensors, laying the foundation for future extensions to other types of sensors.
```

WANGXIAOMIN-HIK committed:
Adjust the order and format of the SensorData content chapters.

WANGXIAOMIN-HIK committed:
Modify the capability definition of SensorData

WANGXIAOMIN-HIK committed:
Adjust the chapter number of Sensor Data Examples

WANGXIAOMIN-HIK committed:
Move Sensor Data Support Attributes from GetServiceCapabilities to GetSupportedMetadata

WANGXIAOMIN-HIK committed:
Remove the definitions of SensorDataOverMQTT and SensorDataFilter, which are part of Metadata

venki5685 commented on wsdl/ver10/schema/metadatastream.xsd:
```diff
@@ -434,9 +434,31 @@ IN NO EVENT WILL THE CORPORATION OR ITS MEMBERS OR THEIR AFFILIATES BE LIABLE FO
+		<xs:sequence>
+			<xs:element name="SensorType" type="xs:string" minOccurs="1" maxOccurs="1"/>
+			<xs:element name="SensorValue" type="xs:float" minOccurs="1" maxOccurs="1"/>
+			<xs:element name="SensorUnit" type="xs:string" minOccurs="0" maxOccurs="1"/>
```
any specific reason Sensor Unit is optional parameter?

WANGXIAOMIN-HIK committed:
**Technical Implementation:**
- Changed `IsProperty="true"` to `IsProperty="false"`
- Changed `minOccurs="0"` to `minOccurs="1"`, making SensorUnit a required parameter

WANGXIAOMIN-HIK commented on doc/Core.xml:
```diff
@@ -6720,6 +6720,84 @@ Event description:
+        
+        <programlisting><![CDATA[Topic: tns1:Device/Sensor/temperature/High
+Event description:
+<tt:MessageDescription IsProperty="true">
```
Thank you for the expert question! Regarding the sensor event types, we have indeed modified sensor events from property events to regular events.

**Modification Rationale:**
Sensor threshold trigger events (such as temperature/humidity High/Low) are essentially **one-time trigger events** that are reported only when sensor values exceed or fall below set thresholds. These events are more appropriately defined as regular events because:
- They are not continuous property changes
- They only occur when thresholds are triggered, with clear trigger timing
- If upper-layer applications need continuous sensor data, they can obtain it through Metadata integration

**Technical Implementation:**
- Changed `IsProperty="true"` to `IsProperty="false"`

WANGXIAOMIN-HIK committed:
Remove the definition of the attribute that the sensor data supports

kieran242 reviewed (commented):
I read over the PR and like the ideas and have raised a couple of questions for clarity?

WANGXIAOMIN-HIK commented on doc/Core.xml:
```diff
@@ -6720,6 +6720,84 @@ Event description:
+          </listitem>
+        </itemizedlist>
+        
+        <para>The following event should be generated with true value when a temperature sensor detects a value exceeding the configured high threshold.</para>
```
Thank you for your thorough review and valuable questions!

**1: Your Question:** "where has this high threshold value been defined on a Sensor?"

**Our Answer:**
Your question about sensor threshold definition is very critical. Indeed, sensor events need to be triggered based on certain thresholds. In current implementations, these thresholds are typically configured through the following methods:

- Device Application Hardcoding: Threshold values are hardcoded in device applications
- Vendor Proprietary Protocols: Configuration through device vendor's proprietary protocols
- Device Web Interface: Manual configuration through device's web management interface

We recognize the importance of clearly defining sensor event threshold configuration in the ONVIF standard. In the future, we will add related sensor event threshold configuration interfaces and mechanisms to the ONVIF standard to improve interoperability and configuration standardization, ensuring unified sensor threshold management across devices from different vendors.

**2: Your Question:** "Are all these sensors attached to a camera device or external to the device relaying the Metadata? i,e. is a Camera or NVR taking metadata from multiple device sources externally and raising metadata events like a relay or performing a triggered action based on the Metadata?"

**Our Answer:**
Our design supports multiple sensor integration modes. ONVIF devices (such as Camera or NVR) can support multiple external sensors connected via direct connection:

- Gateway Mode: ONVIF devices primarily act as gateway roles, directly forwarding data received from external sensors as metadata to upper-layer applications
- Event Trigger Mode: ONVIF devices can also report sensor data as ONVIF events when preset thresholds are reached, implementing triggered actions based on sensor data.

jflevesque-genetec commented on doc/Core.xml:
```diff
@@ -6720,6 +6720,84 @@ Event description:
+            <para>tns1:Device/Sensor/[SensorType] - Generic sensor event topic</para>
+          </listitem>
+          <listitem>
+            <para>tns1:Device/Sensor/temperature/High - Temperature too high event</para>
```
I believe all topics should be Pascal cased so topics are standard. So it would be better to change them to
```
tns1:Device/Sensor/Temperature/High
```

WANGXIAOMIN-HIK committed:
1. Topic Naming Convention Adjustment
2. MetadataStream Structure Optimization
3. Naming Simplification
4. Unit Description Enhancement

WANGXIAOMIN-HIK commented:
We would like to express our sincere gratitude to all the experts who provided valuable feedback and suggestions during the ONVIF Canada conference. Based on the discussions held during the conference, we have implemented the following four key modifications to the ONVIF sensor data standard:

#### 1. Topic Naming Convention Adjustment
**Modification**: Standardized Topic naming convention by capitalizing the first letter
- `tns1:Device/Sensor/temperature/High` → `tns1:Device/Sensor/Temperature/High`
- `tns1:Device/Sensor/temperature/Low` → `tns1:Device/Sensor/Temperature/Low`
- `tns1:Device/Sensor/humidity/High` → `tns1:Device/Sensor/Humidity/High`
- `tns1:Device/Sensor/humidity/Low` → `tns1:Device/Sensor/Humidity/Low`

**Modified Files**: Core.xml (event topic list, MessageDescription examples, MQTT examples)

#### 2. MetadataStream Structure Optimization
**Modification**: Moved SensorDataStream from Extension2 to be at the same level as VideoAnalytics, PTZ, and Event, elevating its importance in the metadata stream
- Added `<xs:element name="SensorDataStream" type="tt:SensorDataStream"/>` after `<xs:element name="Event" type="tt:EventStream"/>`
- Removed SensorDataStream element from MetadataStreamExtension2

**Modified Files**: metadatastream.xsd

#### 3. Naming Simplification
**Modification**: Simplified sensor data related naming to improve readability and consistency
- `SensorType` → `Type`
- `SensorValue` → `Value`
- `SensorUnit` → `Unit`

**Modified Files**:
- metadatastream.xsd (XML Schema definitions)
- Core.xml (MessageDescription and MQTT examples)
- Analytics.xml (all documentation, XML examples, JSON examples)

#### 4. Unit Description Enhancement
**Modification**: Added clarification for absolute temperature and absolute humidity to eliminate ambiguity
- Temperature unit: `"°C" or "°F" or "K" (absolute temperature)`
- Humidity unit: `"%RH" (absolute humidity)`

WANGXIAOMIN-HIK committed:
Update the unit descriptions for sensor data, adding detailed information about temperature and humidity units.

WANGXIAOMIN-HIK committed:
Changes:
1. Schema Update: Add `SensorID` attribute to the `SensorData` complex type in `metadatastream.xsd`
2. Documentation Update: Update related examples in `Analytics.xml` and `Core.xml` to demonstrate multi-sensor data usage scenarios

WANGXIAOMIN-HIK committed:
Remove the 'Type' field definition from 'SensorData'.

WANGXIAOMIN-HIK committed:
Remove Unit, expand Type to include RelativeHumidity, AbsoluteHumidity

WANGXIAOMIN-HIK committed:
Correct the error in the RelativeHumidity example, changing 348.65 to 65.2.

WANGXIAOMIN-HIK committed:
Add unit descriptions for temperature and humidity:
Temperature unit: Kelvin
absolute humidity unit: g/m³
relative humidity unit: %

WANGXIAOMIN-HIK committed:
Change the unit of Absolute Humidity to g/m³

kieran242 reviewed (approved):
@WANGXIAOMIN-HIK  I will be happy to approve this

WANGXIAOMIN-HIK committed:
Refactor sensor data examples, remove unnecessary nested structures, and update element definitions

WANGXIAOMIN-HIK committed:
Add SensorDataFilter

WANGXIAOMIN-HIK committed:
Add the configuration of sensormetadata filtering, SensorDataFilter

WANGXIAOMIN-HIK committed:
Adjust the node position of SensorData

kieran242 reviewed (approved):
@WANGXIAOMIN-HIK I have re-reviewed all your latest changes and still happy to approve.

WANGXIAOMIN-HIK committed:
Increase the detection threshold in the alarm event

WANGXIAOMIN-HIK committed:
Merge branch 'development' of https://github.com/onvif/specs into video/Multi-dimensional-Sensor-Data

WANGXIAOMIN-HIK committed:
Fix the XML structure in Core.xml by adding the missing closing tags.
